File: petties-web/src/services/emrService.ts
import api from './api/client'

export interface EmrRecord {
    id: string
    petId: string
    bookingId?: string
    bookingCode?: string
    staffId: string
    clinicId: string
    clinicName: string
    staffName: string
    petName: string
    petSpecies: string
    petBreed: string
    ownerName?: string
    subjective?: string
    objective?: string
    assessment: string
    plan: string
    notes?: string
    weightKg?: number
    temperatureC?: number
    heartRate?: number
    bcs?: number
    prescriptions: Prescription[]
    images: EmrImage[]
    examinationDate: string
    reExaminationDate?: string
    createdAt: string
    updatedAt?: string
    isLocked?: boolean
}

export interface Prescription {
    medicineName: string
    // ===== New schema fields (preferred) =====
    timesOfDay?: Array<'sang' | 'trua' | 'chieu'>
    beforeAfterMeal?: 'BEFORE_MEAL' | 'AFTER_MEAL' | 'WITH_MEAL' | 'NONE'
    frequencyNote?: string
    durationDays?: number
    instructions?: string

    // ===== Legacy fields (backward-compat only) =====
    dosage?: string
    frequency?: string
}

export interface EmrImage {
    url: string
    description?: string
}

export interface CreateEmrRequest {
    petId: string
    bookingId?: string
    subjective?: string
    objective?: string
    assessment: string
    plan: string
    notes?: string
    weightKg?: number
    temperatureC?: number
    heartRate?: number
    bcs?: number
    prescriptions?: Prescription[]
    images?: EmrImage[]
    reExaminationDate?: string
    examinationDate?: string
    aiDiagnosisContext?: AiDiagnosisContext
}

export interface AiDiagnosisContextPrescription {
    medicine_name?: string
    medicineName?: string
    times_of_day?: Array<'sang' | 'trua' | 'chieu'>
    timesOfDay?: Array<'sang' | 'trua' | 'chieu'>
    before_after_meal?: 'BEFORE_MEAL' | 'AFTER_MEAL' | 'WITH_MEAL' | 'NONE'
    beforeAfterMeal?: 'BEFORE_MEAL' | 'AFTER_MEAL' | 'WITH_MEAL' | 'NONE'
    frequency_note?: string
    frequencyNote?: string
    duration_days?: number
    durationDays?: number
    instructions?: string
    source?: string
    source_detail?: string
    sourceDetail?: string

    // Legacy (for reading old context)
    dosage?: string
    frequency?: string
}

export interface AiDiagnosisContext {
    request_id?: string
    requestId?: string
    selected_diagnosis_code?: string
    selectedDiagnosisCode?: string
    selected_diagnosis_label?: string
    selectedDiagnosisLabel?: string
    suggested_prescriptions?: AiDiagnosisContextPrescription[]
    suggestedPrescriptions?: AiDiagnosisContextPrescription[]
    generated_at?: string
    generatedAt?: string
}

export interface UploadResponse {
    url: string
    publicId: string
    format?: string
    width?: number
    height?: number
    bytes?: number
}

/**
 * EMR Service for Web
 */
export const emrService = {
    /**
     * Get EMR records for a pet
     */
    async getEmrsByPetId(petId: string): Promise<EmrRecord[]> {
        const response = await api.get<EmrRecord[]>(`/emr/pet/${petId}`)
        return response.data
    },

    /**
     * Get EMR by ID
     */
    async getEmrById(emrId: string): Promise<EmrRecord> {
        const response = await api.get<EmrRecord>(`/emr/${emrId}`)
        return response.data
    },

    /**
     * Get EMR by booking ID
     */
    async getEmrByBookingId(bookingId: string): Promise<EmrRecord> {
        const response = await api.get<EmrRecord>(`/emr/booking/${bookingId}`)
        return response.data
    },

    /**
     * Create a new EMR record (Staff only)
     */
    async createEmr(request: CreateEmrRequest): Promise<EmrRecord> {
        const response = await api.post<EmrRecord>('/emr', request)
        return response.data
    },

    /**
     * Update EMR record (Staff only)
     */
    async updateEmr(emrId: string, request: CreateEmrRequest): Promise<EmrRecord> {
        const response = await api.put<EmrRecord>(`/emr/${emrId}`, request)
        return response.data
    },

    /**
     * Upload EMR image
     */
    async uploadEmrImage(file: File, description?: string): Promise<UploadResponse> {
        const formData = new FormData()
        formData.append('file', file)
        if (description) {
            formData.append('description', description)
        }
        const response = await api.post<UploadResponse>('/emr/upload-image', formData, {
            headers: {
                'Content-Type': 'multipart/form-data',
            },
        })
        return response.data
    },
}

export default emrService

File: petties_mobile/lib/data/services/emr_service.dart
import 'package:dio/dio.dart';
import '../models/emr.dart';
import 'api_client.dart';

/// EMR Service - Handle EMR API calls
class EmrService {
  final ApiClient _apiClient;

  EmrService([ApiClient? apiClient]) : _apiClient = apiClient ?? ApiClient.instance;

  /// Get EMR records for a specific pet
  Future<List<EmrRecord>> getEmrsByPetId(String petId) async {
    try {
      final response = await _apiClient.get('/emr/pet/$petId');
      final List<dynamic> data = response.data;
      return data.map((json) => EmrRecord.fromJson(json)).toList();
    } catch (e) {
      rethrow;
    }
  }

  /// Get EMR by ID
  Future<EmrRecord> getEmrById(String emrId) async {
    try {
      final response = await _apiClient.get('/emr/$emrId');
      return EmrRecord.fromJson(response.data);
    } catch (e) {
      rethrow;
    }
  }

  /// Get EMR by Booking ID (returns null if not found)
  Future<EmrRecord?> getEmrByBookingId(String bookingId) async {
    try {
      final response = await _apiClient.get('/emr/booking/$bookingId');
      return EmrRecord.fromJson(response.data);
    } catch (e) {
      // Return null if EMR not found for this booking
      return null;
    }
  }

  /// Create a new EMR record (Vet only)
  Future<EmrRecord> createEmr(CreateEmrRequest request) async {
    try {
      final response = await _apiClient.post(
        '/emr',
        data: request.toJson(),
      );
      return EmrRecord.fromJson(response.data);
    } catch (e) {
      rethrow;
    }
  }

  /// Update EMR record (Vet only)
  Future<EmrRecord> updateEmr(String emrId, CreateEmrRequest request) async {
    try {
      final response = await _apiClient.put(
        '/emr/$emrId',
        data: request.toJson(),
      );
      return EmrRecord.fromJson(response.data);
    } on DioException catch (e) {
      // Extract user-friendly message from API response
      final message = e.response?.data?['message'] ?? 'Lỗi khi cập nhật bệnh án';
      throw Exception(message);
    } catch (e) {
      rethrow;
    }
  }
  /// Upload EMR image
  Future<String> uploadImage(String filePath) async {
    try {
      final formData = FormData.fromMap({
        'file': await MultipartFile.fromFile(filePath),
      });

      final response = await _apiClient.post(
        '/emr/upload-image',
        data: formData,
      );
      
      return response.data['url'];
    } catch (e) {
      rethrow;
    }
  }
}
